### PR TITLE
[rush] Refactor plugin APIs to leverage rush-sdk

### PR DESCRIPTION
**Pull request by octogonz**
## Summary

- Remove the `typeof TheClass` workarounds from `RushSession` since **@rushstack/rush-sdk** enables us to access these classes directly.

- I'm also promoting some APIs to `@public` that have been tagged as `@beta` for a very long time.

- Fix issues flagged by API Extractor 

## Details

We accepted some API engineering debt in order to get the plugin PR merged.  This PR is cleaning up those APIs using the new **@rushstack/rush-sdk** approach, and also fixing API Extractor warnings in the API reports.

👉**NOTE:** Our focus right now is to get the publish feature shipped so the community can start providing feedback about it.  The entire plugin feature is designated as "experimental," and the associated APIs are all marked as `@beta`, so we can continue iterating on these APIs in future releases.  It's okay if some of the APIs aren't fully baked yet.

## How it was tested

Everything compiles.

### CODE REVIEW

File: common/reviews/api/rush-lib.api.md
Comment (octogonz) [inline common/reviews/api/rush-lib.api.md:757]: The `WebClient` API seems very specific to Rush's internals.  Also, it is coupled with the `node-fetch` typings (which are now somewhat controversial since the maintainer made a strange decision to drop support for CommonJS).

We should either make this `@internal` or refactor it to be more general purpose.  I'm marking it as `@beta` only to unblock the release.

File: apps/rush-lib/src/api/RushConfiguration.ts
Comment (iclanton) [inline apps/rush-lib/src/api/RushConfiguration.ts:1733]:
```suggestion
  public get _rushPluginsConfiguration(): RushPluginsConfiguration {
```

File: libraries/rush-sdk/src/index.ts
Comment (iclanton) [inline libraries/rush-sdk/src/index.ts:50]: Add a note about why this isn't being set by the "`// SCENARIO 2:  ...`" comment.

File: apps/rush-lib/src/logic/CredentialCache.ts
Comment (iclanton) [inline apps/rush-lib/src/logic/CredentialCache.ts:43]: Why are you removing the `IDisposable`?

File: common/reviews/api/rush-lib.api.md
Comment (iclanton) [reply common/reviews/api/rush-lib.api.md:757]: What about moving it to node-core-library and putting together our own typings for anything we're reexporting from `node-fetch`? We can then just stick the user agent in the constructor.

File: apps/rush-lib/src/logic/CredentialCache.ts
Comment (octogonz) [reply apps/rush-lib/src/logic/CredentialCache.ts:43]: If we want to make an `IDisposable` public API, it should probably go in `node-core-library`, and should go together with some kind of contract. For example maybe there is a helper API similar to the C# `using (new DisposableThing()) {`.  Or maybe `IDisposable` objects get registered with a manager that takes care of disposing them automatically when the session goes out of scope.

In its current form, `IDisposable` didn't seem to serve any useful purpose besides implying that whoever constructs `CredentialCache` has a responsibility to call `dispose()`.  But is that even the design we want?  For example, maybe it would be better for plugins to request `CredentialCache` from `RushSession` or a `BuildCacheProviderBase` class.  And then Rush itself can take care of disposing `CredentialCache`.  And maybe the `CredentialCache` constructor should be internal.

File: common/reviews/api/rush-lib.api.md
Comment (octogonz) [reply common/reviews/api/rush-lib.api.md:757]: **(Out of scope for this PR)**

That's not a bad idea.  The Rush Stack projects do relatively few HTTP client requests, so we actually haven't put a lot of thought into the right way.  There are several options:

- Node.js exposes native APIs for this, but they're different from the browser runtime, and (last I checked) they are not even promisified
- The standard is `node-fetch` (which ponyfills the browser standard `fetch()`), but as I mentioned that project looks a bit wacky right now.
- The other super popular option is [Axios](https://www.npmjs.com/package/axios) which is a uniform solution for browser and Node.js, and adds [several nice features](https://blog.logrocket.com/axios-or-fetch-api/) (cancellation, timeouts, download progress bars, etc) that go beyond `fetch()`.  The main tradeoff is that it's another dependency 18k minified, 63k for the debug bundle.
- Lastly, if you are writing Node.js services then you'll be using a framework like [Fastify](https://www.fastify.io/) or [Hapi](https://hapi.dev/) which have their own integrated client APIs

`node-core-library` aims to provide a centralized toolkit that enables developers to perform common operations, using best practices, without having to think through these choices themselves.  So for example, if we decided that `fetch()` is the best practice, I agree that it would make sense for `node-core-library` to provide a readymade wrapper for it because it is a tiny API and standardized.  Whereas if we choose one of the other 3 options above, I think we'd simply write docs somewhere saying "we recommend to use that thing."

File: libraries/rush-sdk/src/index.ts
Comment (octogonz) [reply libraries/rush-sdk/src/index.ts:50]: There's an interesting question of whether it would EVER be valid to load two instances of `rush-lib` simultaneously.

For example, suppose a plugin imports some library that has its own bundled `rush-lib` (scenario 1+2).  Or suppose that a tool built with `rush-lib` imports another tool that also has its own `rush-lib` (scenario 2+2). 

- With the code snippet above, whoever loads `rush-lib` first would win.  
- Whereas with the code snippet deleted, then we load two copies side-by-side.

Comment (octogonz) [reply libraries/rush-sdk/src/index.ts:50]: Maybe the safest policy is:
- If `rushLibModule !== undefined` then simply report an error.  If that error is blocking a valid use case, let them file a bug and explain it to us.  :-)

File: rush-plugins/rush-azure-storage-build-cache-plugin/src/AzureStorageBuildCacheProvider.ts
Comment (chengcyber) [inline rush-plugins/rush-azure-storage-build-cache-plugin/src/AzureStorageBuildCacheProvider.ts:10]: Since All API comes from `rush-sdk`, it seems `RushSession` related stuff can be totally removed in this file now.

File: common/reviews/api/rush-lib.api.md
Comment (iclanton) [inline common/reviews/api/rush-lib.api.md:242]: I don't think this should be considered a stable API, as things will disappear from it as we graduate experiments.

File: common/reviews/api/rush-lib.api.md
Comment (iclanton) [inline common/reviews/api/rush-lib.api.md:634]: There's a lot of random stuff in here. I think this should probably be `@beta` so we can clean it up.

File: apps/rush-lib/src/logic/CredentialCache.ts
Comment (iclanton) [reply apps/rush-lib/src/logic/CredentialCache.ts:43]: Fair enough.

File: libraries/rush-sdk/src/index.ts
Comment (iclanton) [reply libraries/rush-sdk/src/index.ts:50]: I think we should just throw an error.

**Review by iclanton — APPROVED** [review]

File: apps/rush-lib/src/api/PackageJsonEditor.ts
Comment (Faithfinder) [inline apps/rush-lib/src/api/PackageJsonEditor.ts:12]: Oh! I know it's not entirely relevant to this PR, but having that as `const` enum doesn't allow it to be used from JS scripts. I've been meaning to open an issue about that.

I.e `packageJsonEditor.addOrUpdateDependency(dep.name, dep.version, DependencyType.Peer);` fails at runtime because `DependencyType.Peer` is `undefined`

Comment (dmichon-msft) [reply apps/rush-lib/src/api/PackageJsonEditor.ts:12]: Exporting const enums is bad practice in general. We should either replace this with a standard enum or a union of literals.

File: libraries/rush-sdk/src/index.ts
Comment (dmichon-msft) [inline libraries/rush-sdk/src/index.ts:7]:
```suggestion
const RUSH_LIB_NAME: '@microsoft/rush-lib' = '@microsoft/rush-lib';
```

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3BuildCacheProvider.ts
Comment (dmichon-msft) [inline rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3BuildCacheProvider.ts:34]: Checking if the global environment variable is set to disable cache writes should be done at a higher level than the specific cache provider, since the environment variable doesn't specify the cache provider. It's useful to be able to block cache writes for the local cache as well.

File: rush-plugins/rush-azure-storage-build-cache-plugin/src/AzureStorageBuildCacheProvider.ts
Comment (dmichon-msft) [inline rush-plugins/rush-azure-storage-build-cache-plugin/src/AzureStorageBuildCacheProvider.ts:6]: It's still better to use type-only imports wherever possible, and to avoid dependencies on types that declare private members.

File: rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3BuildCacheProvider.ts
Comment (dmichon-msft) [reply rush-plugins/rush-amazon-s3-build-cache-plugin/src/AmazonS3BuildCacheProvider.ts:34]: Out of scope for this PR. Handled in !3044

File: apps/rush-lib/src/api/RushConfiguration.ts
Comment (octogonz) [reply apps/rush-lib/src/api/RushConfiguration.ts:1733]: Fixed

File: libraries/rush-sdk/src/index.ts
Comment (octogonz) [reply libraries/rush-sdk/src/index.ts:50]: I thought about this some more.  Imagine this pretty reasonable seeming case:

- `my-tool` depends on `my-library`, and provides `"rush-lib"` as a regular dependency
- `my-tool` imports `rush-sdk`
- `my-library` also imports `rush-sdk`, and possibly installs `"rush-lib"` as a dev dependency for its Jest tests

In this situation, it seems like we can't do much better than "whoever loads `rush-lib` first decides the instance."

To allow that, I've reverted my change above and added some code comments.  When we implement Scenario 3 we can add more validation.

File: rush-plugins/rush-azure-storage-build-cache-plugin/src/AzureStorageBuildCacheProvider.ts
Comment (octogonz) [reply rush-plugins/rush-azure-storage-build-cache-plugin/src/AzureStorageBuildCacheProvider.ts:10]: Let's keep it as a formalism, because generally we DO want to move plugin context onto the session and move away from global state.  As one example, I suggested that maybe `CredentialCache` should eventually be created+disposed by Rush and obtained via `RushSession`.

File: common/reviews/api/rush-lib.api.md
Comment (octogonz) [reply common/reviews/api/rush-lib.api.md:757]: @iclanton and I chatted about this yesterday and decided to temporarily copy+paste `WebClient` to avoid adding it to the `rush-lib` API surface, since we're sure wherever it ends up, it won't be there.

File: apps/rush-lib/src/api/PackageJsonEditor.ts
Comment (octogonz) [reply apps/rush-lib/src/api/PackageJsonEditor.ts:12]: Good catch -- fixed

File: common/reviews/api/rush-lib.api.md
Comment (octogonz) [reply common/reviews/api/rush-lib.api.md:609]: Good point, fixed

File: common/reviews/api/rush-lib.api.md
Comment (octogonz) [reply common/reviews/api/rush-lib.api.md:241]: Fixed

File: libraries/rush-sdk/src/index.ts
Comment (octogonz) [reply libraries/rush-sdk/src/index.ts:7]: Not fixing -- this seems more verbose without serving any useful purpose

**Pull request merged**
